src/hooks/useDataApi.ts

```diff
@@ -0,0 +1,95 @@
+import { useEffect, useReducer } from 'react';
+import axios, { AxiosResponse, AxiosError } from 'axios';
+
+import fetchData from '../utils/fetchData';
+
+type State = {
+  loading: boolean;
+  data?: AxiosResponse['data'];
+  status?: AxiosResponse['status'];
+  statusText?: AxiosResponse['statusText'];
+  headers?: AxiosResponse['headers'];
+  error?: AxiosError;
+};
+
+enum ActionType {
+  INIT = 'INIT',
+  SUCCESS = 'SUCCESS',
+  ERROR = 'ERROR',
+}
+
+type Action =
+  | { type: ActionType.INIT }
+  | { type: ActionType.SUCCESS; response?: AxiosResponse }
+  | { type: ActionType.ERROR; error: AxiosError };
+
+// eslint-disable-next-line consistent-return
+const reducer = (state: State, action: Action): State => {
+  // eslint-disable-next-line default-case
+  switch (action.type) {
+    case ActionType.INIT:
+      return {
+        loading: true,
+      };
+    case ActionType.SUCCESS:
+      return {
+        loading: false,
+        data: action.response && action.response.data,
+        status: action.response && action.response.status,
+        statusText: action.response && action.response.statusText,
+        headers: action.response && action.response.headers,
+      };
+    case ActionType.ERROR:
+      return {
+        loading: false,
+        status: action.error.response && action.error.response.status,
+        statusText: action.error.response && action.error.response.statusText,
+        headers: action.error.response && action.error.response.headers,
+        error: action.error,
+      };
+  }
+};
+
+const useDataApi = (url?: string): State => {
+  const [state, dispatch] = useReducer(reducer, { loading: !!url });
+
+  useEffect(() => {
+    // we don't require a URL, we just don't need data anymore
+    // assume succes with no data
+    if (!url) {
+      dispatch({ type: ActionType.SUCCESS });
+      return;
+    }
+
+    dispatch({ type: ActionType.INIT });
+
+    // variables to handle cancellation
+    const source = axios.CancelToken.source();
+    let didCancel = false;
```

Comment:
I'm wondering whether you need that flag as you could check whether the thrown error is a cancelation with `axios.isCancel(thrown)`. Unless the promise can be resolved as a success even with cancelation?

Reply:
In an ideal world I would have wanted `source` to be able to tell me if it had been cancelled (through something like `source.canceled` for example) but I don't think this is possible.
I did this because I remembered doing something similar in previous code, but using native fetch. In this case you have 2 levels of async, one for getting the response, and one for parsing the JSON (`fetch()` and then `response.json()`) and I was worried that if the cancellation (due to the component unmounting) happened during the 2nd step, the AbortController would not be involved anymore and I wouldn't be able to catch it.
I'm not sure how this happens in axios logic though.
Not sure if I'm making sense 😬 
So since I'm not sure how to test my theory I leant on the safe side here by using my own flag.

Reply:
So if you want me to use `isCancel` I'm fine with it if you think the way axios handles it is OK.

Reply:
Makes sense. I don't mind either way.